### PR TITLE
Eddystone TLM

### PR DESCRIPTION
**Pull request by ioarun**
Added Eddystone TLM feature to the library.
Usage can be found in the examples folder.

### CODE REVIEW

File: EddystoneBeacon.cpp [hunk 1]
```diff
@@ -7,16 +7,17 @@
...
-  "http://www.",
-  "https://www.",
+  "http://www.",//for http://www.
+  "https://www", //for .
   "http://",
   "https://",
-  "urn:uuid:"
+  "urn:uuid:",
+  "http//localhost:" //additional url prefix for connecting to localhost.
 };
```
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: This addition is not up to spec, see: https://github.com/google/eddystone/tree/master/eddystone-url#url-scheme-prefix

File: EddystoneBeacon.cpp [hunk 2]
```diff
@@ -7,16 +7,17 @@
...
 
 static const char* EDDYSTONE_URL_BEACON_SUFFIX_SUBSTITUTIONS[] = {
-  ".com/",
-  ".org/",
+  ".c",  //com
+  ".o",  //org
```
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: These changes are not up to spec, see: https://github.com/google/eddystone/tree/master/eddystone-url#eddystone-url-http-url-encoding

File: EddystoneBeacon.cpp [hunk 3]
```diff
@@ -28,7 +29,8 @@ static const char* EDDYSTONE_URL_BEACON_SUFFIX_SUBSTITUTIONS[] = {
   ".net",
   ".info",
   ".biz",
-  ".gov"
+  ".gov",
+  ""
```
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: This changes are not up to spec, see: https://github.com/google/eddystone/tree/master/eddystone-url#eddystone-url-http-url-encoding

File: EddystoneBeacon.cpp [hunk 4]
```diff
@@ -36,6 +38,8 @@ EddystoneBeacon::EddystoneBeacon(unsigned char req, unsigned char rdy, unsigned
   _bleService("feaa"),
   _bleCharacteristic("feab", BLERead | BLEBroadcast, MAX_SERVICE_DATA_SIZE)
 {
+  this->setAdvertisedServiceUuid(this->_bleService.uuid());
+
```
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: Why was this moved up?

File: EddystoneBeacon.cpp [hunk 5]
```diff
@@ -76,6 +78,18 @@ void EddystoneBeacon::begin(char power, const char* uri) {
 
   this->_bleCharacteristic.broadcast();
 }
+/***Following is the code for broadcasting TLM ****/
+/***Implementation in the examples folder***/
+void EddystoneBeacon::begin(const unsigned char* temp, unsigned char len){
```
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: A nicer API would be to specify the battery voltage and temperature as numbers, instead of raw TLM data.

**Comment by sandeepmistry** [comment]
@arun271 thank you for submitting!

However these changes cannot be merged, although they are functional, they do not meet the Eddystone specification. 

When in TLM mode, the "Advertising PDU count" and "Time since power-on or reboot" will need to be updated as per https://github.com/google/eddystone/tree/master/eddystone-tlm#frame-specification.

This could be achieved in the ``EddystoneBeacon::loop()`` method, by calculating the PDU count and uptime via mills(). Then updating an internal TLM data buffer.

**Comment by arun271** [comment]
@sandeepmistry 
Sir,Thanks a lot for the reviews.I have come up with few changes and tried to make the Eddystone TLM API easier to use.These changes include temperature,PDU count and uptime.
In Eddystone.cpp,added the following:
1.
uint8_t buffer[14]= {
  0x20,
  0x00,
  0x00,
  0x00,
  0x00,
  0x80,
  0x00,
  0x00,
  0x00,
  0x00,
  0x00,
  0x00,
  0x00,
  0x00
}; //global
unsigned char len = 14; //global
2.
void EddystoneBeacon::tlm_begin(){

  this->_bleCharacteristic.setValue(buffer, len);
  BLEPeripheral::begin();
  this->_bleCharacteristic.broadcast();
}
3.
void EddystoneBeacon::tempLoop(int analogPin) {
  this->poll();
  pdu_count++;
  /** temperature **/
  pinMode(analogPin,INPUT);
  int a = analogRead(analogPin);
  resistance=(float)(1023-a)*10000/a; //get the resistance of the sensor;
  temperature=1/(log(resistance/10000)/B+1/298.15)-273.15;//convert to temperature via datasheet
  buffer_1[4] = ((float2fix(temperature))&0x0000FF00)>>8; //LSB
  buffer_1[5] = (float2fix(temperature))&0x000000FF;  //MSB
  //temperature

  unsigned long val = millis()/100;
  ulong2adv(buffer,10, val);  //UP-TIME SINCE POWER ON OR REBOOT
  ulong2adv(buffer,6, pdu_count);  //PDU COUNT

  _bleCharacteristic.setValue(buffer, len);
}
4.
void EddystoneBeacon::ulong2adv(uint8_t* adv, int off, unsigned long val) {
  off+=3;
  for (int i = 0; i < 4; ++i) {
    uint8_t x = val & 0xff;
    adv[off] = x;
    val >>= 8;
    off--;
  }
}

5.Corresponding changes will be done for defining these new methods in Eddystone.h
6.The usage example is here : [https://github.com/arun271/arduino-BLEPeripheral/blob/eddystone/examples/Eddystone/EddystoneTLM](url)

Except for the battery-voltage data, the frame is complete I guess.Please review the changes and let me know what more can be added.
Thanks!

**Comment by arun271** [comment]
@sandeepmistry 
In case above link doesn't open,I have attached the usage code below:
`// Import libraries (EddystoneBeacon depends on SPI)
#include <SPI.h>
#include <EddystoneBeacon.h>

// define pins (varies per shield/board)
//
//   Adafruit Bluefruit LE   10, 2, 9
//   Blend                    9, 8, UNUSED
//   Blend Micro              6, 7, 4
//   RBL BLE Shield           9, 8, UNUSED

#define EDDYSTONE_BEACON_REQ   9
#define EDDYSTONE_BEACON_RDY   8
#define EDDYSTONE_BEACON_RST   7

int analogPin = 0;

EddystoneBeacon eddystoneBeacon = EddystoneBeacon(EDDYSTONE_BEACON_REQ, EDDYSTONE_BEACON_RDY, EDDYSTONE_BEACON_RST);

void setup() {
  pinMode(analogPin,INPUT);
  Serial.begin(57600);

  delay(1000);

  eddystoneBeacon.tlm_begin();

}

void loop() {
  eddystoneBeacon.tempLoop(analogPin);
  
}`

File: EddystoneBeacon.cpp (hunk 4 above)
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: Ping ^

File: EddystoneBeacon.cpp [hunk 6]
```diff
@@ -76,6 +76,33 @@ void EddystoneBeacon::begin(char power, const char* uri) {
 
   this->_bleCharacteristic.broadcast();
 }
+/***Following is the code for broadcasting TLM ****/
+/***Implementation in the examples folder***/
+uint8_t buffer[14]= {
```
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: This should be a member var declared in ``` EddystoneBeacon.h```

File: EddystoneBeacon.cpp [hunk 7]
```diff
@@ -76,6 +76,33 @@ void EddystoneBeacon::begin(char power, const char* uri) {
...
+  0x00,
+  0x00,
+  0x00,
+  0x00,
+  0x00,
+  0x00,
+  0x00
+}; //global
+unsigned char len = 14; //global
+void EddystoneBeacon::tlm_begin(){
```
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: I'd prefer an API like: ```void EddystoneBeacon::begin(int temperature, int batteryLevel)```

File: EddystoneBeacon.cpp [hunk 8]
```diff
@@ -117,3 +144,35 @@ unsigned char EddystoneBeacon::compressURI(const char* uri, char *compressedUri,
 void EddystoneBeacon::loop() {
   this->poll();
 }
+/* loop() for TLM below:**/
+float temperature;
+int B=3975;                  //B value of the thermistor
+float resistance;
+unsigned long pdu_count = 0;
+void EddystoneBeacon::tempLoop(int analogPin) {
```
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: I don't think this needs be here, the existing ```::loop()``` function can be used.

Also, we should have a ```setTLMData(int temperature, int batteryLevel);``` API. The temperature sensor code can go in an example sketch instead.

File: EddystoneBeacon.cpp (hunk 7 above)
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: Maybe we don't even need a buffer, just splitting vars for temperature, battery level, adv count, boot time etc.

File: EddystoneBeacon.cpp [hunk 9]
```diff
@@ -117,3 +144,35 @@ unsigned char EddystoneBeacon::compressURI(const char* uri, char *compressedUri,
...
+  int a = analogRead(analogPin);
+  resistance=(float)(1023-a)*10000/a; //get the resistance of the sensor;
+  temperature=1/(log(resistance/10000)/B+1/298.15)-273.15;//convert to temperature via datasheet&nbsp;;
+  buffer_1[4] = ((float2fix(temperature))&0x0000FF00)>>8; //LSB
+  buffer_1[5] = (float2fix(temperature))&0x000000FF;  //MSB
+  //temperature
+  
+  unsigned long val = millis()/100;
+  ulong2adv(buffer,10, val);  //UP-TIME SINCE POWER ON OR REBOOT
+  ulong2adv(buffer,6, pdu_count);  //PDU COUNT
```
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: I think the PDU count is inaccurate, since the advertisement interval is 100ms, it can be estimate via time from the call to begin in millis / 100.

File: EddystoneBeacon.cpp [hunk 10]
```diff
@@ -117,3 +144,35 @@ unsigned char EddystoneBeacon::compressURI(const char* uri, char *compressedUri,
...
+
+  _bleCharacteristic.setValue(buffer, len);
+}
+void EddystoneBeacon::ulong2adv(uint8_t* adv, int off, unsigned long val) {
```
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: This can probably be replaced with a ```memcpy```.

File: examples/EddyTLM.ino [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+#define EDDYSTONE_BEACON_REQ   9
+#define EDDYSTONE_BEACON_RDY   8
+#define EDDYSTONE_BEACON_RST   7
+
+int analogPin = 0;
+int resetPin = 7;
+int button = 2;
+float temp = 0.0;
+
+unsigned char buffer_1[14] = {
```
Comment (sandeepmistry) [inline examples/EddyTLM.ino]: These buffer's seem unused?

File: examples/Eddystone/EddystoneTLM [hunk 12]
```diff
@@ -0,0 +1,35 @@
+// Import libraries (EddystoneBeacon depends on SPI)
+#include <SPI.h>
+#include <EddystoneBeacon.h>
+
```
Comment (sandeepmistry) [inline examples/Eddystone/EddystoneTLM]: Can this file be deleted?

File: examples/EddyTLM.ino [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+  delay(1000);
+
+  eddystoneBeacon.begin(ptr1,len);
+
+}
+
+void loop() {
+  Serial.println("main loop");
+  eddystoneBeacon.loop();
+  if(digitalRead(button) == HIGH){
```
Comment (sandeepmistry) [inline examples/EddyTLM.ino]: This seems like a pretty complex example, we should start with a really basic TLM example, that just reads and update temperature and battery level from sensors.

**Comment by sandeepmistry** [comment]
@arun271 overall great progress!

See my new comments, feel free to reply to them on Github before making the changes.

File: EddystoneBeacon.cpp (hunk 7 above)
Comment (arun271) [inline EddystoneBeacon.cpp]: Having declared and defined `uint8_t buffer[14]` in` Eddystone.h` I changed `tlm_begin()` to the following:
void EddystoneBeacon::begin(float temp, float bLevel){
  
  this->_bleCharacteristic.setValue(buffer, len);

  BLEPeripheral::begin();

  this->_bleCharacteristic.broadcast();

}

I have added `setTLMData(float temp,float bLevel)` to set the TLM values.

File: EddystoneBeacon.cpp (hunk 8 above)
Comment (arun271) [inline EddystoneBeacon.cpp]: Deleted `tempLoop()` and added `setTLMData(float temp,float bLevel)`

void EddystoneBeacon::setTLMData(float temp, float bLevel) {
  buffer[2] = ((float2fix(bLevel))&0x0000FF00)>>8; //LSB
  buffer[3] = (float2fix(bLevel))&0x000000FF;  //MSB
  buffer[4] = ((float2fix(temp))&0x0000FF00)>>8; //LSB
  buffer[5] = (float2fix(temp))&0x000000FF;  //MSB
   val = millis()/100;
  ulong2adv(buffer,10, val);  //UP-TIME SINCE POWER ON OR REBOOT
  ulong2adv(buffer,6, pdu_count);  //PDU COUNT

  _bleCharacteristic.setValue(buffer, len);
}

Comment (arun271) [inline EddystoneBeacon.cpp]: *One more thing.I forgot to add `pdu_count++` in the above code segment.

File: EddystoneBeacon.cpp (hunk 9 above)
Comment (arun271) [inline EddystoneBeacon.cpp]: Sorry Sir,I am not sure if I fully understood.But I have added `pdu_count++` inside `setTLMData()`.Please help me here.

File: examples/EddyTLM.ino (hunk 11 above)
Comment (arun271) [inline examples/EddyTLM.ino]: This was an old example `/examples/EddyTLM.ino`.

File: examples/Eddystone/EddystoneTLM (hunk 12 above)
Comment (arun271) [inline examples/Eddystone/EddystoneTLM]: Actually this was the modified(better and simpler) one. I want to keep this example (`/examples/EddystonTLM`)  and still better and simpler example code would be following which I recently wrote for the above mentioned changes.

// Import libraries (EddystoneBeacon depends on SPI)
#include <SPI.h>
#include <EddystoneBeacon.h>

// define pins (varies per shield/board)
//
//   Adafruit Bluefruit LE   10, 2, 9
//   Blend                    9, 8, UNUSED
//   Blend Micro              6, 7, 4
//   RBL BLE Shield           9, 8, UNUSED

#define EDDYSTONE_BEACON_REQ   9
#define EDDYSTONE_BEACON_RDY   8
#define EDDYSTONE_BEACON_RST   7

int tempPin = 0;
int batteryPin = 1;
float temp = 0.0;
float bLevel = 0.0;
float resistance = 0.0;
int B=3975; 

EddystoneBeacon eddystoneBeacon = EddystoneBeacon(EDDYSTONE_BEACON_REQ, EDDYSTONE_BEACON_RDY, EDDYSTONE_BEACON_RST);

void setup() {
  pinMode(tempPin,INPUT);
  pinMode(batteryPin,INPUT);
  Serial.begin(57600);
  
  delay(1000);
  
  eddystoneBeacon.begin(temp,bLevel);

}

void loop() {
  int a = analogRead(tempPin);
  resistance=(float)(1023-a)*10000/a; //get the resistance of the sensor;
  temp=1/(log(resistance/10000)/B+1/298.15)-273.15;//convert to temperature via datasheet
  eddystoneBeacon.setTLMData(temp,bLevel);
  eddystoneBeacon.loop();
  
}

File: examples/EddyTLM.ino (hunk 13 above)
Comment (arun271) [inline examples/EddyTLM.ino]: Yes Sir, this one was `/examples/EddyTLM.ino`. I think I have already deleted this example.Please take a look at the newer example that I have written in the previous comment.

File: EddystoneBeacon.cpp (hunk 4 above)
Comment (arun271) [inline EddystoneBeacon.cpp]: That is a mistake.I was trying something and then forgot to delete that line.

File: EddystoneBeacon.cpp (hunk 10 above)
Comment (arun271) [inline EddystoneBeacon.cpp]: @sandeepmistry 
I have removed `ulong2adv()` and then used memcpy as follows:
1.a new array called `newData[12]` will hold new values from the sensors.
2.using `memcpy()`,copied the `newData` to `buffer` from index 2 to 13.
3.Then setTLMData() looks as below.

void EddystoneBeacon::setTLMData(float temp, float bLevel) {
  pdu_count++;
  uint8_t newData[12];
  newData[0] = ((float2fix(bLevel))&0x0000FF00)>>8; //LSB
  newData[1] = (float2fix(bLevel))&0x000000FF;  //MSB
  newData[2] = ((float2fix(temp))&0x0000FF00)>>8; //LSB
  newData[3] = (float2fix(temp))&0x000000FF;  //MSB
  int off = 7;
  for (int i = 0; i < 4; ++i) {
    uint8_t bval = pdu_count & 0xff;
    newData[off] = bval;
    val >>= 8;
    off--;
  }
  val = millis()/100;
  off = 11;
  for (int i = 0; i < 4; ++i) {
    uint8_t bval = val & 0xff;
    newData[off] = bval;
    val >>= 8;
    off--;
  }
  memcpy(buffer + 2,newData, 12);
  _bleCharacteristic.setValue(buffer, len);
}

Comment (arun271) [inline EddystoneBeacon.cpp]: I was wondering why `pdu_count` should be inaccurate because, it will be update after every setTLMData() call from the example sketch i.e everytime before the frame is broadcasted.

Comment (arun271) [inline EddystoneBeacon.cpp]: Or should I put it inside `loop()`.

File: EddystoneBeacon.cpp (hunk 9 above)
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: My understanding is PDU count is number of advertising frames. If it's based on time setTLMData is called, I think this will be incorrect. Default advertising interval 100ms. So 10 frames will be advertised every 1 second.

File: examples/EddyTLM.ino (hunk 11 above)
Comment (sandeepmistry) [inline examples/EddyTLM.ino]: Please delete then.

File: examples/Eddystone/EddystoneTLM (hunk 12 above)
Comment (sandeepmistry) [inline examples/Eddystone/EddystoneTLM]: but it doesn't have the .ino file extension.

File: EddystoneBeacon.cpp (hunk 8 above)
Comment (sandeepmistry) [inline EddystoneBeacon.cpp]: We should avoid global variables. I still don't think those are needed same with the buffers. They can be local variables.

**Comment by sandeepmistry** [comment]
Closing this for now due to lack of activity.
